### PR TITLE
8252377: Incorrect encoding for EC AlgorithmIdentifier

### PR DESCRIPTION
**Pull request by haimaychao**
This change fixes the DER encoding for ECDSA AlgorithmIdentifier to omit the parameters field instead of encoding a Null tag.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8252377](https://bugs.openjdk.java.net/browse/JDK-8252377): Incorrect encoding for EC AlgorithmIdentifier

### Reviewers
 * [Weijun Wang](https://openjdk.java.net/census#weijun) (@wangweij - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/312/head:pull/312`
`$ git checkout pull/312`

### CODE REVIEW

**Comment by wangweij** [comment]
I don't quite understand what the test is for. The bug is about encoding but the test seems to be decoding the certificates. Does the test fail before this fix and succeed after it?

**Comment by haimaychao** [comment]
This is because the encoding of Algorithm Identifier incorrectly adds two NULL tags to the parameters field in the canned certificate. (There are two Algorithm Identifiers in the cert, with each NULL tag containing two bytes:  tag + length.) I use the length of an encoded certificate (x509Cert.getEncoded().length) to verify that the certificate contains an extra 4 bytes to hold the two NULL tags. Therefore, the certificate without the fix should be 4 bytes (5 bytes if one byte alignment is applied) longer in length than the certificate with the fix.

**Comment by wangweij** [comment]
But the certificates included in the test are static and if one day we mistakenly add the NULL back it will not be detected by the test. My idea is to generate a cert on the fly inside the test so that can check the `derEncode` method in *this* JDK. As for checking if the NULL is there, you can try looking at the `DerUtils::shouldNotExist` method with "021" and "11" as arguments.

File: src/java.base/share/classes/sun/security/x509/AlgorithmId.java [hunk 1]
```diff
@@ -203,7 +203,13 @@ public void derEncode (OutputStream out) throws IOException {
...
-                    || algid.equals(ed25519_oid)) {
+                    || algid.equals(ed25519_oid)
+                    || algid.equals(x448_oid)
+                    || algid.equals(x25519_oid)
+                    || algid.equals(SHA224withECDSA_oid)
+                    || algid.equals(SHA256withECDSA_oid)
+                    || algid.equals(SHA384withECDSA_oid)
+                    || algid.equals(SHA512withECDSA_oid)) {
                 // RFC 4055 3.3: when an RSASSA-PSS key does not require
                 // parameter validation, field is absent.
```
Comment (wangweij) [inline src/java.base/share/classes/sun/security/x509/AlgorithmId.java:214]: Would you like to add more comments on other RFCs here?

Comment (haimaychao) [reply src/java.base/share/classes/sun/security/x509/AlgorithmId.java:214]: Added the comments for RFC reference.

**Review by wangweij — APPROVED** [review]

**Pull request closed without merging**